### PR TITLE
vote model and API endpoint added

### PR DESCRIPTION
**Pull request by Ali-D-Akbar**
# Blog App Backend:

Description: Vote model added with two endpoints, upvote and downvote.

Installation Instructions:
1 .Change directory to `webblog/` and make a virtual environment by running `virtualenv env`
2 .Open terminal and start virtual environment by running `source ./env/bin/activate`
3 .run `pip install -r requirements.txt --no-index --find-links file:///tmp/packages` to make requirements
4. run `python manage.py runserver` to start the django server

### CODE REVIEW

File: blog/migrations/0001_initial.py [hunk 1]
```diff
@@ -22,6 +20,7 @@ class Migration(migrations.Migration):
                 ('description', models.TextField(max_length=1500)),
                 ('created', models.DateTimeField(auto_now_add=True)),
                 ('image', models.FileField(blank=True, null=True, upload_to='')),
+                ('votes', models.IntegerField(default=0)),
```
Comment (zubair-arbi) [inline blog/migrations/0001_initial.py]: We should never change a migrations like this. Remove all the changes from this migration file

File: blog/migrations/0003_auto_20190926_1412.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
+# Generated by Django 2.2.5 on 2019-09-26 14:12
+
+from django.conf import settings
+from django.db import migrations
+
+
+class Migration(migrations.Migration):
```
Comment (zubair-arbi) [inline blog/migrations/0003_auto_20190926_1412.py]: Combine this with the migration `0002`

File: blog/migrations/0001_initial.py (hunk 1 above)
Comment (Ali-D-Akbar) [reply blog/migrations/0001_initial.py]: done

File: blog/migrations/0003_auto_20190926_1412.py (hunk 2 above)
Comment (Ali-D-Akbar) [reply blog/migrations/0003_auto_20190926_1412.py]: done

File: blog/models.py [hunk 3]
```diff
@@ -3,25 +3,42 @@
...
-    description = models.TextField(max_length=1500)
+    description = models.TextField(max_length=50000)
     created = models.DateTimeField(auto_now_add=True)
     owner = models.ForeignKey('auth.User', related_name='blog', on_delete=models.CASCADE, null=True)
     image = models.FileField(blank=True, null=True)
 
+    @property
+    def votes(self):
+        upvotes = UserVote.objects.filter(blog=self, vote_type='up').count()
+        downvotes = UserVote.objects.filter(blog=self, vote_type='down').count()
```
Comment (zubair-arbi) [inline blog/models.py]: Try to combine these 2 queries in order to get the calculated votes. Also declare the class variables for up vote type and down vote type and use them in the project.

File: blog/models.py [hunk 4]
```diff
@@ -3,25 +3,42 @@
...
+        return upvotes - downvotes
+
+    def upvote(self, user):
+        vote, created = self.post_votes.get_or_create(user=user, blog=self)
+        if not created and vote.vote_type == 'up':
+            return 'already_voted'
```
Comment (zubair-arbi) [inline blog/models.py]: This status should be verbose. I believe we are using it as a message to display on the frontend, in this case this message should be something like `You have already up voted this blog post.`

File: blog/models.py [hunk 5]
```diff
@@ -3,25 +3,42 @@
...
+
+        vote.vote_type = 'up'
+        vote.save()
+        return 'ok'
```
Comment (zubair-arbi) [inline blog/models.py]: `You have successfully up voted this blog post.`

File: blog/models.py [hunk 6]
```diff
@@ -3,25 +3,42 @@
...
+
+    def downvote(self, user):
+        vote, created = self.post_votes.get_or_create(user=user, blog=self)
+        if not created and vote.vote_type == 'down':
+            return 'already_voted'
```
Comment (zubair-arbi) [inline blog/models.py]: update these messages  same as up vote.

File: blog/models.py (hunk 3 above)
Comment (Ali-D-Akbar) [reply blog/models.py]: I looked into Annotation/Aggregation but I can't find a better solution.

Although I've added the VOTE_CHOICE in the Vote Model.

File: blog/models.py (hunk 4 above)
Comment (Ali-D-Akbar) [reply blog/models.py]: Done

File: blog/models.py (hunk 5 above)
Comment (Ali-D-Akbar) [reply blog/models.py]: Done

File: blog/models.py (hunk 6 above)
Comment (Ali-D-Akbar) [reply blog/models.py]: Done

File: blog/models.py (hunk 3 above)
Comment (Ali-D-Akbar) [reply blog/models.py]: I've done this using aggregate.
I'm updating the PR.

**Pull request merged**